src/iam/permhandler/permhandler.cpp

```diff
@@ -0,0 +1,127 @@
+
+RetWithError<StaticString<uuid::cUUIDStrLen>> PermHandler::RegisterInstance(
+    const InstanceIdent& instanceIdent, const Array<FunctionalServicePermissions>& instancePermissions)
+{
+    LockGuard lock(mMutex);
+
+    Error                           err;
+    StaticString<uuid::cUUIDStrLen> uuidSecret;
+
+    Tie(uuidSecret, err) = GetSecretForInstance(instanceIdent);
+    if (err.IsNone()) {
+        LOG_DBG() << "Instance " << instanceIdent << " has been already registered";
+
+        return {uuidSecret, err};
+    }
+
+    uuidSecret = GenerateSecret();
+    err        = AddSecret(uuidSecret, instanceIdent, instancePermissions);
+
+    if (!err.IsNone()) {
+        return {"", err};
+    }
+
+    return {uuidSecret, err};
+}
+
+Error PermHandler::UnregisterInstance(const InstanceIdent& instanceIdent)
+{
+    LockGuard lock(mMutex);
+
+    auto err
+        = mInstancesPerms.Find([&instanceIdent](const auto& item) { return item.mInstanceIdent == instanceIdent; });
+    if (!err.mError.IsNone()) {
+        LOG_WRN() << "Instance " << instanceIdent << " not registered";
+
+        return err.mError;
+    }
+
+    return mInstancesPerms.Remove(err.mValue).mError;
+}
+
+Error PermHandler::GetPermissions(InstanceIdent& instanceIdent, Array<PermKeyValue>& servicePermissions,
+    const String& secretUUID, const String& funcServerID)
+{
+    LockGuard lock(mMutex);
+
+    auto err = mInstancesPerms.Find([&secretUUID](const auto& item) { return secretUUID == item.mSecretUUID; });
+    if (!err.mError.IsNone()) {
+        LOG_ERR() << "secret not found";
+
+        return err.mError;
+    }
+
+    instanceIdent = err.mValue->mInstanceIdent;
+
+    for (const auto& it : err.mValue->mFuncServicePerms) {
+        if (it.mName == funcServerID) {
+            if (it.mPermissions.Size() > servicePermissions.MaxSize()) {
+                return ErrorEnum::eNoMemory;
+            }
+
+            servicePermissions = it.mPermissions;
+
+            return ErrorEnum::eNone;
+        }
+    }
+
+    LOG_ERR() << "permissions for functional server " << funcServerID << " not found";
+
+    return ErrorEnum::eNotFound;
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+Error PermHandler::AddSecret(const String& secretUUID, const InstanceIdent& instanceIdent,
+    const Array<FunctionalServicePermissions>& instancePermissions)
+{
+    if (mInstancesPerms.Size() == mInstancesPerms.MaxSize()) {
+        LOG_ERR() << "Max number of secrets stored has been reached";
+
+        return ErrorEnum::eNoMemory;
+    }
+
+    return mInstancesPerms.PushBack(InstancePermissions {secretUUID, instanceIdent, instancePermissions});
+}
+
+StaticString<uuid::cUUIDStrLen> PermHandler::GenerateSecret()
+{
+    const auto uuid = uuid::CreateUUID();
```

Comment:
Our UUID generator unfortunately doesn't guarantee unique values. So we need to keep check unique and retry mechanism. Similar we have in go implementation.